feat: Initial implementation

philip-paul-mueller opened the pull request:
Initial JaCe code.

This PR introduces a series of very basic functionalities, that will be extended in subsequent PRs.
Most importantly it introduces the basic infrastructure that allows to translate a Jaxpr object into an SDFG.
Furthermore, it also adds the `jace.jit` decorator, which can be used as a replacement for `jax.jit`.
However, a function that was decorated with `jace.jit` remains fully composable with Jax transformation, such as `jax.grad` or `jax.jacfwd`.
But, the functionality is still very basic and `jace.jit`, essentially does not accepts any arguments and only works on CPU.
Nevertheless, there is a cache for caching tracing, translation and compilation of wrapped functions.

Although, this PR introduces the components for the translation, the actual primitive translators are not yet implemented (this commit adds the `ALUTranslator`, but this one was back ported from the prototype for simple tests).

As a last point the tests, located in `tests` are only a first version, that were not included in the review of this PR.
They will be reviewed at a later stage

For more information see the `ROADMAP.md` file.

egparedes reviewed (commented):
A first round of comments. I skipped the review of parts of the code until we further discuss the general approach with @edopao .

Additionally, proper unit tests should be added for the current functionality.

philip-paul-mueller commented on src/jace/translator/jace_subtranslator_interface.py:
```diff
@@ -0,0 +1,148 @@
+    from .jaxpr_translator_driver import JaxprTranslationDriver
+
+
+class JaCeSubTranslatorInterface:
```
I changed it, and I agree it makes the code a bit more readable.
However, I do not like the name because it suggests that it is able to already translate everything.

egparedes commented on src/jace/util/util.py:
```diff
@@ -0,0 +1,30 @@
+from typing import Any
+
+
+def ensure_iterability(
+    x: Any,
+    ign_str: bool = True,
+) -> Iterable[Any]:
+    """Ensures that `x` is iterable.
+
+    By default strings are _not_ considered iterable.
+
+    Args:
+        x:          To test.
+        ign_str:    Ignore that a string is iterabile.
+    """
+    if ign_str and isinstance(x, str):
+        x = [x]
+    elif isinstance(x, Iterable):
+        pass
+    return x
```
Ok, there are two aspects here: one is checking if an object is an iterable instance which is not a string and the second is to wrap an element into a single-element sequence. I was talking only about the first one, in which case the `ign_str` parameter doesn't make sense because it has to be `True` for this function to be useful, since when it is `False` it'd be equivalent to `isinstance(x, Iterable)`. 

So what I proposed was to define the function as a typeguard (https://typing.readthedocs.io/en/latest/spec/narrowing.html#typeguard) for a new `Iterable` subtype (e.g. `NonStringIterable`) representing Iterable instances that are not strings. Something like:

```python
class NonStringIterable(Iterable): ...

def is_non_string_iterable(val: Iterable) -> Typeguard[NonStringIterable]:
    return isinstance(val, Iterable) and not isinstance(val, str)
```
A slightly more _magic_ alternative would be to directly implement a subclashook in the custom Iterable subclass:
```python
class NonStringIterable(Iterable):
    @classmethod
    def __subclasshook__(cls, C):
        return not isinstance(C, str) and super().__subclasshook__(C)
```
For the second aspect, wrapping atomic elements into single-element sequences, you can define a similar function as the current one but again I'd drop the `ign_str` argument since it's currently unused (and most likely it will always be that way), and I'd add the right overloads to accurate express the types, as for example: 

```python
@overload
def as_sequence(value: str) -> Iterable[str]: ...

@overload
def as_sequence(value: Iterable[T]) -> Iterable[T]: ...

@overload
def as_sequence(value: T) -> Iterable[T]: ...

def as_sequence(value: T | Iterable[T]) -> Iterable[T]:
    return [value] if not isinstance(value, Iterable) or isinstance(value, str) else value
```

Note: all the code here is basically untested pseudocode

philip-paul-mueller replied:
Thanks I was not aware of this.
I made an implementation, of your code and it worked, kind of, but you should take a look as it is very likely not fully correct.

egparedes commented on src/jace/translator/sub_translators/__init__.py:
```diff
@@ -0,0 +1,92 @@
+#
+# SPDX-License-Identifier: BSD-3-Clause
+
+"""Module collecting all built-in subtranslators."""
```
Why do we need to distinguish between primitive translator based on who implemented them? Why do we need to delete primitive translators? I don't see the use cases, and certainly not now, so to me this module looks like it's bloated with functionality which is not currently needed.

philip-paul-mueller replied:
Distinguish them would allow that an external user can override a built in primitive.
At initialization, first the externally supplied translators would be created.
Then one would go through the build in, if a subtranslator was already registered for it we know that it came from the outside and was supposed to override the built in and we would skip it.
I think it is quite an elegant solution.

egparedes commented on src/jace/translator/util/jace_translation_memento.py:
```diff
@@ -0,0 +1,76 @@
+
+
+@dataclass(init=True, repr=True, eq=False, frozen=True, kw_only=True, slots=True)
+class JaCeTranslationMemento:
```
What about something straightforward like `SDFGContext`, `TranslationContext` or something along these lines? It would be shorter, simpler to remember and easier to understand. `TranslatedSDFG` is slightly misleading in my opinion since this class contains more stuff than a single SDFG, but does not inherit or extend the SDFG class in any other way.

philip-paul-mueller replied:
I agree that `TranslatedSDFG` is misleading, I have now settled with `TranslatedJaxprSDFG`.
I do not use `*Context` as I decided to not use this class inside the driver, instead I created a different class.
The main motivation reason for this is that in my view both are exceptionally rather different.

egparedes commented on src/jace/translator/jace_subtranslator_interface.py:
```diff
@@ -0,0 +1,148 @@
+        It is required that subclasses calls this method during initialization.
+        """
+
+    def get_handled_primitives(self) -> Collection[str] | str:
```
Again, this should be defined as a property not as getter method, and both this property and the `translate` method should be defined as `abstractmethods` which would help to catch partially undefined subclasses much earlier.  I also think now that the case where the same object implements more than one primitive is just an implementation detail and should not be part of the interface,  so the `primitive` property should be just a single string.

philip-paul-mueller replied:
- I turned it into a property (although I do not like it).
- I was not aware that for a Protocol one has to explicitly declare the methods as abstract.
- I do not think that it is an implementation detail, thus I will keep the signature as it is.

philip-paul-mueller commented on src/jace/translator/jace_subtranslator_interface.py:
```diff
@@ -0,0 +1,148 @@
+        """Initialize the interface.
+
+        It is required that subclasses calls this method during initialization.
+        """
```
As we discussed on Tuesday:
- We should keep the amount of configuration minimal
- We should come up with a different scheme, most likely store the configuration in the driver.

(correct me if I forgot something)

egparedes commented on src/jace/util/jax_helper.py:
```diff
@@ -0,0 +1,201 @@
+
+
+@overload
+def get_jax_var_shape(jax_var: JaCeVar) -> tuple[int | dace.symbol | str, ...] | tuple[()]: ...
```
What does `tuple[()]` mean in this type annotation? This is not needed for representing an empty tuple.

egparedes commented on src/jace/util/jax_helper.py:
```diff
@@ -0,0 +1,201 @@
+
+def get_jax_var_shape(
+    jax_var: jax_core.Atom | JaCeVar,
+) -> tuple[int | dace.symbol | str, ...] | tuple[()]:
```
Same question as above about `tuple[()]`

egparedes commented on tests/test_jaxpr_translator_driver.py:
```diff
@@ -0,0 +1,360 @@
+
+
+@pytest.fixture(scope="module")
+def alloc_driver():
```
Fixtures shouldn't be named as verbs but as substantives, since they are used as proxies for their result.
  
```suggestion
def translation_driver():
```

egparedes reviewed (changes requested):
I've skipped most of the general design discussions for now and only focused on python programming style, so the requested changes are small and shouldn't be hard to implement quickly.

philip-paul-mueller commented on src/jace/translator/jaxpr_translator_driver.py:
```diff
@@ -0,0 +1,1039 @@
+        """
+        return self._ctx.sdfg
+
+    def get_terminal_sdfg_state(self) -> dace.SDFGState:
```
There is only one function, `get_array()` which takes an argument that can not be translated.

philip-paul-mueller commented on src/jace/translator/jaxpr_translator_driver.py:
```diff
@@ -0,0 +1,1039 @@
+        self._ctx.jax_name_map[jax_var] = sdfg_name
+        return self
+
+    def add_reserved_names(
+        self,
+        reserved_names: None | str | Collection[str],
+    ) -> JaxprTranslationDriver:
+        """Adds the names listed in `reserved_names` to the internal list."""
+
+        if reserved_names is None:
+            return self
+        if isinstance(reserved_names, str):
+            reserved_names = [reserved_names]
+        elif isinstance(reserved_names, Collection):
+            pass
+        else:
+            raise TypeError(f"Does not know how to handle the type '{type(reserved_names)}'.")
+        self._reserved_names.update(reserved_names)
+        return self
```
- I hate functions that return `None` i much more like if they return `self` which allows for method chaining.
- It is allowed to pass `None` to indicate that nothing should be added to the set. As you pointed out you can turn it into an empty set.
- Oversight from my side.

philip-paul-mueller commented on src/jace/translator/jaxpr_translator_driver.py:
```diff
@@ -0,0 +1,1253 @@
+from jace.translator import sub_translators as jtsubt, util as jtrutil
+
+
+class JaxprTranslationDriver:
```
At some point we will rename it to builder.

philip-paul-mueller reviewed (commented):
I think I addressed all of your comments.
I reworked the registering of subtranslators, which is now in `jace.translator.managing`.
The solution I found there was the least bad solution.

egparedes commented on src/jace/jax/translation_cache.py:
```diff
@@ -0,0 +1,297 @@
+    fargs: CallArgsDescription
+
+
+class TranslationCache:
```
Why cannot use the standard `__getitem__`, `__setitem__`, `__delitem__` and `__contains__` operators instead of custom function (`get`, `add`, `_evict`, `has`) doing exactly the same?

egparedes commented on src/jace/jax/translation_cache.py:
```diff
@@ -0,0 +1,297 @@
+
+
+@dataclass(init=True, eq=True, frozen=True)
+class _AbstarctCallArgument:
```
Why not naming this just `_AbstractValue` in the same spirit of jax?

philip-paul-mueller replied:
In Jax `_AbstractValue` has a different meaning it is used for the tracer and has as far as I know nothing to do with caching.
Furthermore, an `_AbstractValue` could be anywhere, but we only use it inside the cache to describe the arguments of a call, thus this name is much more appropriate.

egparedes commented on src/jace/jax/translation_cache.py:
```diff
@@ -0,0 +1,297 @@
+    return _TRANSLATION_CACHES[tstage]
+
+
+@dataclass(init=True, eq=True, frozen=True)
```
I would remove the arguments matching the default values.
```suggestion
@dataclass(frozen=True)
```

egparedes commented on src/jace/jax/translation_cache.py:
```diff
@@ -0,0 +1,297 @@
+        raise TypeError(f"Can not make 'an abstract description from '{type(val).__name__}'.")
+
+
+@runtime_checkable
+class _ConcreteCallArgument(Protocol):
+    """Type for encoding a concrete arguments in the cache."""
+
+    @abstractmethod
+    def __hash__(self) -> int:
+        pass
+
+    @abstractmethod
+    def __eq__(self, other: Any) -> bool:
+        pass
```
This protocol is very generic and doesn't define any requirement linked to concrete call arguments so I'd recommend to use a generic name (e.g. `Comparable`) and consider moving it to a more general module to make it available in other parts of Jace.

philip-paul-mueller replied:
I deleted it and uses `Hashable`, however, I think it is less expressive.

egparedes commented on src/jace/jax/translation_cache.py:
```diff
@@ -0,0 +1,297 @@
+        **kwargs: Any,
+    ) -> stages.Stage:
+        # Get the abstract description of the call, that is used as key.
+        key: CachedCallDescription = self._make_call_decscription(*args, **kwargs)
```
I think it  would be cleaner to define a Protocol for stages (e.g. `CachableStage`) which define a method to generate a key for the translation cache  (e.g. `make_cache_key()`) and use it as type hint for `self` here.

philip-paul-mueller replied:
I implemented it, but the cache is on the function call.
Consider the first stage, `JaceWrapped`, which is not cached at all, but `JaceWrapped::lower()` has to be cached.
The same goes for the last stage `JaceCompiled` which is cached, since it is the result of `JaceLowered::compile()`, but it does not need a `make_cache_key()` function.
Thus, I named the protocols `CachedStageTransition`, not that great, but a proper name.

philip-paul-mueller commented on src/jace/jax/api.py:
```diff
@@ -0,0 +1,137 @@
+            i.e. the output of `get_regsitered_primitive_translators()`, are used.
+            After construction changes to the passed `sub_translators` have no effect on the returned object.
+    """
+    if len(kwargs) != 0:
```
I agree, this was actually there, but I removed it.

philip-paul-mueller commented on src/jace/jax/api.py:
```diff
@@ -0,0 +1,137 @@
+        )
+
+    def wrapper(f: Callable) -> stages.JaceWrapped:
+        from jace import jax as stages  # Cyclic import
```
It was a typo.

philip-paul-mueller commented on src/jace/jax/api.py:
```diff
@@ -0,0 +1,137 @@
+from collections.abc import Callable, Mapping
+from typing import TYPE_CHECKING, Any, Literal, overload
+
+import jax as _jax_jax
```
Other parts of the code use this `_jax` convention too.
I switched to it because it was indeed cryptic.
However, according to [PEP8](https://peps.python.org/pep-0008/#package-and-module-names), this is actually signalizes a C/C++ extension module.

egparedes replied:
> However, according to [PEP8](https://peps.python.org/pep-0008/#package-and-module-names), this is actually signalizes a C/C++ extension module.

If you refer to the following paragraph of PEP8, I don't think is related:

> When an extension module written in C or C++ has an accompanying Python module that provides a higher level (e.g. more object oriented) interface, the C/C++ module has a leading underscore (e.g. _socket).

This only explains how a C extension module related to a python module should be named, but doesn't say that this is the only possible case to use an underscore in a module name or in an import alias for a 3rd party library:

philip-paul-mueller commented on src/jace/jax/api.py:
```diff
@@ -0,0 +1,137 @@
+        sub_translators:    Use these subtranslators for the lowering to DaCe.
+
+    Notes:
+        If no subtranslators are specified then the ones that are currently active,
+            i.e. the output of `get_regsitered_primitive_translators()`, are used.
+            After construction changes to the passed `sub_translators` have no effect on the returned object.
```
As I understand it, it can be a list, I already used that thing multiple times.
This allows you to distinguish it.
But I will change that.

philip-paul-mueller commented on src/jace/jax/api.py:
```diff
@@ -0,0 +1,137 @@
+    )
+
+
+def grad(
+    fun: Callable | None = None,
+    /,
+    **kwargs: Any,
+) -> Callable:
+    """Jace wrapper for `jax.grad`.
+
+    Notes:
+        Note we can not put it into a `JaceWrapped` object because in autodiff mode
+            control primitives, such as `if` are allowed, but not in `jit`.
+            Thus there need to be this extra layer.
+    """
+    return _jax_jax.grad(fun, **kwargs)
+
+
+def jacfwd(
+    fun: Callable | None = None,
+    /,
+    **kwargs: Any,
+) -> Callable:
+    """Jace wrapper around `jax.jacfwd`."""
+    return _jax_jax.jacfwd(fun, **kwargs)
+
+
+def jacrev(
+    fun: Callable | None = None,
+    /,
+    **kwargs: Any,
+) -> Callable:
+    """Jace wrapper around `jax.jacrev`."""
+    return _jax_jax.jacrev(fun, **kwargs)
```
No, in fact these is everything that is needed.
Actually the only thing that is missing is a wrapper, that replaces `JAX` with `JaCe` in the docstring and we are done.

However, I have removed them.

philip-paul-mueller commented on src/jace/jax/stages.py:
```diff
@@ -0,0 +1,325 @@
+from typing import Any, Final, TypeAlias
+
+import dace
+import jax as jax_jax
```
I agree.

philip-paul-mueller commented on src/jace/jax/stages.py:
```diff
@@ -0,0 +1,325 @@
+    """
+
+
+"""Map type to pass compiler options to `JaceLowered.compile()`.
+"""
+CompilerOptions: TypeAlias = dict[str, tuple[bool, str]]
```
Actually, this is exactly what Jax does, there it is also a type alias inside the `stages` module.
I also think that adding this class is things making too complicated and is not really needed right now.
So for the sake of reducing lines, we will not do that.

However, I changed the form of the comment.

philip-paul-mueller commented on src/jace/jax/stages.py:
```diff
@@ -0,0 +1,325 @@
+    """
+
+    _fun: Callable
+    _sub_translators: Mapping[str, translator.PrimitiveTranslatorCallable]
```
The only thing that `PrimitiveTranslator` adds is basically the `.primitive` property, that you do not use in any way inside the driver.
The only place where it is important is inside the `register_primitive_subtranslator()` function and nowhere else.
Essentially, it imposes the fewest restrictions.

I do not remember exactly, but I started this rewrite, because MyPy complained about something, that was related to that.

philip-paul-mueller commented on src/jace/jax/stages.py:
```diff
@@ -0,0 +1,325 @@
+from jace.util import dace_helper as jdace
+
+
+class Stage:
+    """A distinct step in the compilation chain, see module description for more.
+
+    The concrete steps are implemented in:
+    - JaceWrapped
+    - JaceLowered
+    - JaceCompiled
+    """
```
This module mimics Jax and in Jax it is a class.
Furthermore, the original one adds some more properties and stuff that we do not support yet, such as pytrees.
Thus I keep it.

philip-paul-mueller replied:
Made it into a union.

philip-paul-mueller commented on src/jace/jax/stages.py:
```diff
@@ -0,0 +1,325 @@
+        #  since it was cached, but it would actually contain an already optimized SDFG, which is not what we want.
+        #  If you think you can remove this line then do it and run `tests/test_decorator.py::test_decorator_sharing`.
+        fsdfg: translator.TranslatedJaxprSDFG = copy.deepcopy(self._trans_sdfg)
+        optimization.jace_optimize(
+            fsdfg, **(self.DEF_COMPILER_OPTIONS if compiler_options is None else compiler_options)
+        )
```
I am not sure, but I want that `lowered.compile()` uses the default options and `lowered.compile({})` compiles uses all options disabled.
Because, this is what `jace_optimize()` does.

I agree that the call `lowered.compile(None)` means "use default is not that good".

philip-paul-mueller commented on src/jace/jax/stages.py:
```diff
@@ -0,0 +1,325 @@
+class JaceLowered(Stage):
+    """Represents the original computation that was lowered to SDFG."""
+
+    DEF_COMPILER_OPTIONS: Final[dict[str, Any]] = {
+        "auto_optimize": True,
+        "simplify": True,
+    }
```
It is an error to annotate something with both `Final` and `ClassVar`.
According to [SO](https://stackoverflow.com/a/66285309) you should only use Final.

philip-paul-mueller commented on src/jace/__init__.py:
```diff
@@ -5,13 +5,24 @@
 
 from __future__ import annotations
 
+import jace.translator.primitive_translators  # noqa: F401  # needed to poulate the internal list of translators.
```
I did it, but it still complains so I added the `noqa` again.
However I kept the `as _` because it is nicely indicating that it is not really used.

egparedes commented on src/jace/jax/translation_cache.py:
```diff
@@ -0,0 +1,290 @@
+
+
+@dataclasses.dataclass(frozen=True)
+class CachedCallDescription:
```
This object represents the information of a call to a specific Stage method (or _action_, as it is sometimes named in this module) with a specific set of arguments. I think the name is misleading because it doesn't mention that it represents a call to a Stage transformation method and instead it mentions _caching_, which is completely unrelated to the content of the object (using this object for caching is just  one possible use).

I would consider using other names like `ActionCallDescription`  or `StageTransformationCall` or something along this line.

philip-paul-mueller replied:
I used `StageTransformationDescription`.

egparedes reviewed (changes requested):
I reviewed again the basic compiler infrastructure. Most of my comments are questions and comments about python style, but the overall design and implementation looks correct to me.

I would expect that this part doesn't need further review and should be ready for being merged after addressing the commented issues and doing a full pass over the comments and docstrings to fix typos, too long line lengths (read the coding conventions again), pieces out-of-sync with the surrounding code and formatting. Remember that both docstrings and source comments should be written as correct English texts expected to be read by other developers.  Except for the very minor lightweight markups for sections used in the Google Python Style docstring conventions, they should avoid formatting marks (e.g. markdown literal quotes, weird indentations, ...) and should use regular writing marks (e.g. capitalization, punctuation, ...)  for emphasis or clarification.

philip-paul-mueller commented on src/jace/__init__.py:
```diff
@@ -5,13 +5,35 @@
+#  some aggressive [type promotion](https://jax.readthedocs.io/en/latest/type_promotion.html).
+#  However, in this case we will have problems when we call the SDFG, for some reasons
+#  `CompiledSDFG` does not work in that case correctly, thus we enable it now globally.
+_jax.config.update("jax_enable_x64", True)
```
You are right it is not the best place.
The reason is that by default Jax traces everything as float32, even if the input is float64 unless this option is in place.
This then leads to errors if it is called in DaCe, Jax solves this issue by performing some hidden casting step.
However, currently this setting is essentially needed, but I fully agree with our point, so I moved it to where the actual tracing happens, further it will not affect the outside.

philip-paul-mueller commented on src/jace/jax/stages.py:
```diff
@@ -0,0 +1,295 @@
+    """
+
+    _fun: Callable
+    _sub_translators: dict[str, translator.PrimitiveTranslator]
```
"Subtranslator" is the original name, it will be slowly replaced by primitive translator, but it will take some time.

philip-paul-mueller commented on src/jace/jax/stages.py:
```diff
@@ -0,0 +1,295 @@
+
+        # Currently we do not allow memory order beside `C_CONTIGUOUS`.
+        #  This is the best place to check for it.
+        if not all((not util.is_array(arg)) or arg.flags["C_CONTIGUOUS"] for arg in args):
```
Actually the memory layout does not influence lowering directly.
By default DaCe uses continuous C order for all arrays which is fine and as long as you have only such input there is no issue.
One way to support such inputs, which was used by the prototype, is to add a layer of arrays with the strides of the input, i.e. arguments to `lower()`, during construction of the SDFG.

I spend several days of my life figuring that out, so the check will remain, however, it should probably go into `run_jax_sdfg()`.
But there is no good point.

Another, more complex one, is to make them symbolic and determine them at call time.

philip-paul-mueller commented on src/jace/jax/stages.py:
```diff
@@ -0,0 +1,295 @@
+        self,
+        trans_sdfg: translator.TranslatedJaxprSDFG,
+    ) -> None:
+        """Constructs the lowered object."""
```
The style guide says that both are okay, but it must be consistent.
See https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

philip-paul-mueller commented on src/jace/jax/stages.py:
```diff
@@ -0,0 +1,295 @@
+        trans_sdfg: translator.TranslatedJaxprSDFG,
+    ) -> None:
+        """Constructs the lowered object."""
+        if not trans_sdfg.is_finalized:
+            raise ValueError("The translated SDFG must be finalized.")
+        if trans_sdfg.inp_names is None:
+            raise ValueError("Input names must be defined.")
+        if trans_sdfg.out_names is None:
+            raise ValueError("Output names must be defined.")
```
Actually it is enough to only check for finalized.

This method you mentioned already exists, but it also validates the SDFG.

philip-paul-mueller commented on src/jace/jax/stages.py:
```diff
@@ -0,0 +1,295 @@
+        #  Since all optimizations works in place, we would violate this assumption.
+        tsdfg: translator.TranslatedJaxprSDFG = copy.deepcopy(self._trans_sdfg)
+
+        # Must be the same as in `_make_call_description()`!
+        options = optimization.DEFAULT_OPTIMIZATIONS | (compiler_options or {})
```
This is a good idea.

philip-paul-mueller commented on src/jace/jax/translation_cache.py:
```diff
@@ -0,0 +1,290 @@
+        only the structure of them are important, this is similar to the tracers used in Jax.
+    - Concrete description: Here one caches on the actual value of the argument,
+        which is similar to static arguments in Jax.
+        The only restriction is that they are hash able.
```
Since I put the constructor arguments in the `__init__`, which is allowed by the convention, I moved them to the `attributes` instead.

philip-paul-mueller commented on src/jace/jax/translation_cache.py:
```diff
@@ -0,0 +1,290 @@
+
+    The decorator will call the annotated function only if the call is not stored inside the cache.
+    The key to look up the call in the cache is computed by `self._make_call_description()`.
+    For this the stage must be derived from `CachingStage`.
```
Actually I found a better solution, https://mypy.readthedocs.io/en/stable/generics.html#declaring-decorators.
In this one MyPy is also able to figuring out the correct return types.
It is a bit of an overblown however.

philip-paul-mueller replied:
I now found a solution, it does not look good, but it detects all the errors.

philip-paul-mueller commented on src/jace/jax/translation_cache.py:
```diff
@@ -0,0 +1,290 @@
+    fargs: CallArgsDescription
+
+
+class TranslationCache:
```
According to the [google doc example](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) it is allowed to put them into `__init__`, as long as it is everywhere the same.

egparedes reviewed (changes requested):
Some more general comments about the translation pipeline.

philip-paul-mueller commented on src/jace/translator/jaxpr_translator_driver.py:
```diff
@@ -0,0 +1,754 @@
+            #  But currently the best we can do, is forbid it!
+            raise NotImplementedError(
+                "You have disabled 'x64' support in Jax, which interferes with the calling of the SDFG. "
+                "SDFG generated in this way will fail to call."
```
I fully removed that part, it is now handled upstream.

philip-paul-mueller commented on src/jace/translator/__init__.py:
```diff
@@ -0,0 +1,32 @@
+# JaCe - JAX Just-In-Time compilation using DaCe (Data Centric Parallel Programming)
```
Todo:

- Consistent code style and documentation (e.g. JaCe casing, writing style, punctuation)
-  ~Consistent and reasonable use of __slots__~
- ~Merge translator/managing.py and translator/primitive_translator.py~
- ~Rename "sub_translators" to "primitive_translators"~
- Rename Driver to Builder/Translator?
- Discuss refactor of TranslatedJaxSDFG as JaxSDFG and TranslationContext = (JaxSDFG, states, ...)

egparedes commented on src/jace/translator/primitive_translators/alu_translator.py:
```diff
@@ -0,0 +1,292 @@
+                assert len(dims_to_bcast) == 0
+                i_memlet = dace.Memlet.simple(in_var_names[i], "0")
+            else:  # Array: We may have to broadcast
+                inputs_: list[str] = []
+                for dim, (map_var, _) in enumerate(tskl_map_ranges):
+                    if dim in dims_to_bcast:
+                        inputs_.append("0")
+                    else:
+                        inputs_.append(map_var)
```
This is basically a list comprehension, right? 
```suggestion
                inputs_: list[str] = [
                    "0" if dim in dims_to_bcast else map_var
                    for dim, (map_var, _) in enumerate(tskl_map_ranges)
                ]
```

egparedes commented on src/jace/translator/primitive_translators/alu_translator.py:
```diff
@@ -0,0 +1,292 @@
+                    else:
+                        inputs_.append(map_var)
+                i_memlet = dace.Memlet.simple(in_var_names[i], ", ".join(inputs_))
+                del inputs_
```
Why?

egparedes commented on src/jace/translator/primitive_translators/alu_translator.py:
```diff
@@ -0,0 +1,292 @@
+
+        # Generate the Memlets for the input.
+        for i, dims_to_bcast in zip(range(len(in_var_names)), [dims_to_bcastl, dims_to_bcastr]):
+            if in_var_names[i] is None:  # Literal: No input needed.
+                tskl_inputs.append((None, None))
+                continue
```
I just don't understand this code...

egparedes commented on src/jace/translator/primitive_translators/alu_translator.py:
```diff
@@ -0,0 +1,292 @@
+                tskl_code,
+            )
+            for in_var, (in_connector, in_memlet) in zip(in_var_names, tskl_inputs, strict=False):
+                if in_var is None:  # So access node for literal
+                    continue
+                eqn_state.add_edge(
+                    eqn_state.add_read(in_var),
+                    None,
+                    tskl_tasklet,
+                    in_connector,
+                    in_memlet,
+                )
```
```suggestion
                if in_var is not None:  # So access node for literal
                    eqn_state.add_edge(
                        eqn_state.add_read(in_var),
                        None,
                        tskl_tasklet,
                        in_connector,
                        in_memlet,
                    )
```

egparedes commented on src/jace/translator/primitive_translators/alu_translator.py:
```diff
@@ -0,0 +1,292 @@
+        return t_code
+
+
+def _list_to_dict(inp: Sequence[tuple[None | Any, Any]]) -> dict[Any, Any]:
```
Just `list_to_dict()` is misleading. I would make it more explicit if possible, like:
```suggestion
def _list_to_filtered_dict(inp: Sequence[tuple[None | Any, Any]]) -> dict[Any, Any]:
```

egparedes commented on src/jace/translator/translated_jaxpr_sdfg.py:
```diff
@@ -0,0 +1,85 @@
+from jace import util
+
+
+class TranslatedJaxprSDFG:
```
Why not just `JaxprSDFG`? The translated part seems a bit redundant to me...

philip-paul-mueller replied:
I decided to keep it that way, mostly because in docstrings I refered to it as "translated SDFG" it is nice to refer to it.

egparedes reviewed (changes requested):
The overall state of the PR has improved quite a lot in my opinion, kudos for all the hard refactoring cleanup work!

This review doesn't include comments for `util` and `tests`. I will add another review just for them.

philip-paul-mueller commented on src/jace/api.py:
```diff
@@ -0,0 +1,87 @@
+
+
+def jit(
+    fun: Callable | None = None,
```
I added a todo, but I do not think that this is particular useful, as Jax also does not care about that.

philip-paul-mueller commented on src/jace/optimization.py:
```diff
@@ -0,0 +1,82 @@
+    Note:
+        Its main job is to exists that we have something that we can call in the tool chain.
+    """
+    if not kwargs:
+        return
+
+    # Unpack the arguments, defaults are such that no optimization is done.
+    simplify = kwargs.get("simplify", False)
+    auto_optimize = kwargs.get("auto_optimize", False)
+
+    if simplify:
+        tsdfg.sdfg.simplify()
+
+    if auto_optimize:
+        pass
```
Thanks for noticing the problem with the early exit.

philip-paul-mueller commented on src/jace/stages.py:
```diff
@@ -0,0 +1,333 @@
+import jax as _jax
+
+from jace import optimization, translator, util
+from jace.optimization import CompilerOptions
```
It is needed because it is exported for compatibility with Jax.

philip-paul-mueller commented on src/jace/stages.py:
```diff
@@ -0,0 +1,333 @@
+        #  aggressive [type promotion](https://jax.readthedocs.io/en/latest/type_promotion.html).
+        #  However, in this case we will have problems when we call the SDFG, for some reasons
+        #  `CompiledSDFG` does not work in that case correctly, thus we enable it for the tracing.
+        with _jax.experimental.enable_x64():
```
In an earlier version there was a warning here, but it become so annoying.
Also disabling is not really useful, so I keep it this way, because it the most useful and practical version.

philip-paul-mueller commented on src/jace/stages.py:
```diff
@@ -0,0 +1,333 @@
+            the default arguments.
+        """
+        # We **must** deepcopy before we do any optimization, because all optimizations are in
+        #  place, however, to properly cache stages, they have to be immutable.
```
The stage object.

philip-paul-mueller commented on src/jace/stages.py:
```diff
@@ -0,0 +1,333 @@
+            return self._translated_sdfg
+        raise ValueError(f"Unknown dialect '{dialect}'.")
+
+    def as_html(self, filename: str | None = None) -> None:
```
I see you point, but these kind of functions in Jax are all named `as_something()` thus we keep them.

egparedes replied:
But the key is that this function is not the same kind of function because it's not returning a value but running a side effect. The current `as_html()` name doesn't make sense and I still think it should be changed.

philip-paul-mueller replied:
You are right, I now named it plain `view()`.

philip-paul-mueller commented on src/jace/stages.py:
```diff
@@ -0,0 +1,333 @@
+        """Runs the `view()` method of the underlying SDFG."""
+        self.compiler_ir().sdfg.view(filename=filename, verbose=False)
+
+    def as_sdfg(self) -> dace.SDFG:
```
This is named in the spirit of the Jax Stage members.

philip-paul-mueller commented on src/jace/stages.py:
```diff
@@ -0,0 +1,333 @@
+        - Handle pytrees.
+    """
+
+    _csdfg: dace_helper.CompiledSDFG  # The compiled SDFG object.
+    _inp_names: tuple[str, ...]  # Name of all input arguments.
+    _out_names: tuple[str, ...]  # Name of all output arguments.
```
There is no public view of the attributes, there are just the arguments.
But I get what you meant and I agree.

philip-paul-mueller commented on src/jace/translator/__init__.py:
```diff
@@ -0,0 +1,38 @@
+"""Subpackage containing all the code related to the Jaxpr to SDFG translation.
+
+The concrete primitive translators that ships with JaCe are inside the `primitive_translators`
+subpackage.
```
According to https://stackoverflow.com/a/7948672 package, or subpackage is the correct term.

egparedes replied:
Yes, you're right. I didn't realize the docstring was inide an `__init__.py`. 

By the way, you could have directly checked the proper definition of `package` in the official documentation (https://docs.python.org/3/glossary.html#term-package)  instead of stackoverflow.

philip-paul-mueller commented on src/jace/translator/jaxpr_translator_builder.py:
```diff
@@ -0,0 +1,828 @@
+from dace import data as ddata, properties as dprop
+from jax import core as jax_core
+
+
+if TYPE_CHECKING:
+    from jace import translator
+
+from jace import util
```
I tried it and it does it not.
It probably wrote it in this order, because I tought ruff would put it in the correct order, but it does not look like that.
So much for automation.

philip-paul-mueller commented on src/jace/translator/jaxpr_translator_builder.py:
```diff
@@ -0,0 +1,828 @@
+        elif allow_fail:
+            return None
+        else:
+            KeyError(f"The Jax variable '{jax_var}' was never registered.")
```
Thanks a lot.

philip-paul-mueller commented on src/jace/translator/jaxpr_translator_builder.py:
```diff
@@ -0,0 +1,828 @@
+        if arg_name in util.FORBIDDEN_SDFG_VAR_NAMES:
+            raise ValueError(f"add_array({arg}): The proposed name '{arg_name}', is forbidden.")
+
+        self._ctx.sdfg.add_array(
+            name=arg_name,
+            shape=shape,
+            strides=strides,
+            offset=offset,
+            storage=storage,
+            dtype=dtype,
+            transient=as_transient,
+        )
+
+        if update_var_mapping:
+            try:
+                # If the mapping fails, remove the variable from the SDFG.
+                self.add_jax_name_mapping(jax_var=arg, sdfg_name=arg_name)
+            except:
+                del self._ctx.sdfg.arrays[arg_name]
+                raise
```
What if adding an array fails?

philip-paul-mueller commented on src/jace/translator/jaxpr_translator_builder.py:
```diff
@@ -0,0 +1,828 @@
+        )
+
+        # Determine the new (tentative) terminal state of the SDFG we are building.
+        if new_sdfg_term_state is None:
+            if eqn_state is not self._ctx.terminal_state:
+                raise RuntimeError("Inconsistent terminal state was detected.")
+            new_sdfg_term_state = eqn_state
+        if not self._ctx.validate():
+            raise RuntimeError("Detected an invalid SDFG under construction.")
```
Yes, because it essentially checks if the primitive translator mixed up.
You have to keep in mind that this, although it look as if it is very deep inside Jace, is actually user facing code.

philip-paul-mueller commented on src/jace/translator/jaxpr_translator_builder.py:
```diff
@@ -0,0 +1,828 @@
+        for expectedSDFGName, jax_var in zip(out_var_names, eqn.outvars, strict=True):
+            mapped_sdfg_name = self.map_jax_var_to_sdfg(jax_var)
+            if mapped_sdfg_name != expectedSDFGName:
+                raise ValueError(
```
Actually this part will be removed.

philip-paul-mueller commented on src/jace/translator/jaxpr_translator_builder.py:
```diff
@@ -0,0 +1,828 @@
+    @inp_names.setter
+    def inp_names(self, inp_names: tuple[str, ...]) -> None:
+        if len(inp_names) == 0:
+            raise dace.sdfg.InvalidSDFGError(
```
The intention was that this is actually the process of building an SDFG, thus in my view an invalid SDFG error is more appropriate than a `ValueError`, but I see your point.

philip-paul-mueller commented on src/jace/optimization.py:
```diff
@@ -0,0 +1,82 @@
+
+from typing import TYPE_CHECKING, Final, TypedDict
+
+from typing_extensions import Unpack
```
Technically yes.
I think that the rule ignores the `typing` and `typing_extension` modules, because otherwise all import code would look like:
```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from typing import THE_REST
```

This was also a reason why I was against the TCH* rules, because in the end it is inconsistent, because anything you import from typing, except `TYPE_CHECKING` (for funny reasons), should be, by its very nature, inside this block!

philip-paul-mueller commented on src/jace/translator/jaxpr_translator_builder.py:
```diff
@@ -0,0 +1,828 @@
+    def _create_constants(
+        self,
+        jaxpr: jax_core.ClosedJaxpr,
+    ) -> Sequence[str]:
```
Okay, but why?

egparedes commented on src/jace/util/compiling.py:
```diff
@@ -0,0 +1,146 @@
+# JaCe - JAX Just-In-Time compilation using DaCe (Data Centric Parallel Programming)
```
`compiling` is a misleading name for this module, as even the docstring states that it does more than just compilation. I'm also not sure if this module belongs to `.util` and not as regular top-level module. Why not just rename to something like `jace.execution`?

philip-paul-mueller replied:
Execution is misleading in the same way as it does more than just execute the SDFG.

philip-paul-mueller replied:
I have now moved them to `dace_helper.py`.

egparedes commented on src/jace/util/compiling.py:
```diff
@@ -0,0 +1,146 @@
+        of the input `TranslatedJaxprSDFG`.
+    """
+    if any(  # We do not support the DaCe return mechanism
+        arrname.startswith("__return")
+        for arrname in tsdfg.sdfg.arrays.keys()  # noqa: SIM118  # we can not use `in` because we are also interested in `__return_`!
```
Rename `arrname` to either `arr_name` (ugly) or just `name` to follow coding conventions.

egparedes commented on src/jace/util/compiling.py:
```diff
@@ -0,0 +1,146 @@
+    csdfg: dace_helper.CompiledSDFG,
+    inp_names: Sequence[str],
+    out_names: Sequence[str],
+    cargs: Sequence[Any],
+    ckwargs: Mapping[str, Any],
```
As mentioned in previous reviews, this naming style is not very pythonic and not very coherent with the other parameter names in this function. I think either `args` & `kwags` or `call_args` & `call_kwargs` would be more appropriate and standard.

egparedes commented on src/jace/util/compiling.py:
```diff
@@ -0,0 +1,146 @@
+        raise NotImplementedError("No kwargs are supported yet.")
+    if len(inp_names) != len(cargs):
+        raise RuntimeError("Wrong number of arguments.")
+    if len(sdfg.free_symbols) != 0:  # This is a simplification that makes our life simple.
```
```suggestion
    if sdfg.free_symbols:  # This is a simplification that makes our life simple.
```

egparedes commented on src/jace/util/compiling.py:
```diff
@@ -0,0 +1,146 @@
+        csdfg(**call_args)
+
+    # Handling the output (pytrees are missing)
+    if len(out_names) == 0:
```
```suggestion
    if not out_names:
```

egparedes commented on src/jace/util/compiling.py:
```diff
@@ -0,0 +1,146 @@
+
+    # Handling the output (pytrees are missing)
+    if len(out_names) == 0:
+        return None
```
This is contradicting the return type annotations. Is the type checker complaining here? If not, why?

philip-paul-mueller replied:
MyPy does not complain.

egparedes commented on src/jace/util/compiling.py:
```diff
@@ -0,0 +1,146 @@
+    if len(out_names) == 0:
+        return None
+    ret_val: tuple[Any] = tuple(call_args[out_name] for out_name in out_names)
+    if len(out_names) == 1:
+        return ret_val[0]
+    return ret_val
```
```suggestion
    return ret_val[0] if len(out_names) == 1 else return ret_val
```

egparedes commented on src/jace/util/traits.py:
```diff
@@ -0,0 +1,137 @@
+    While a return value of `True` guarantees a jaxified object, `False` does not proof the
+    contrary. See also `jace.util.is_jaceified()` to tests if something is a JaCe object.
+    """
+    jaxifyed_types = (
```
```suggestion
    jaxified_types = (
```

egparedes commented on src/jace/util/traits.py:
```diff
@@ -0,0 +1,137 @@
+    Similar to `is_jaxified` but for JaCe objects.
+    """
+
+    if util.is_jaxified(obj):
+        return False
+    return isinstance(obj, stages.JaCeWrapped)
```
Why only checking if is an instance of `JaceWrapped` and not of the other stages?
```suggestion
    return not util.is_jaxified(obj) and isinstance(obj, stages.JaCeWrapped)
```

philip-paul-mueller replied:
I deleted it.

egparedes commented on src/jace/util/traits.py:
```diff
@@ -0,0 +1,137 @@
+
+def is_jaxified(
+    obj: Any,
+) -> TypeGuard[jax_core.Primitive | jax_src.pjit.JitWrapped | jax_xe.PjitFunction]:
```
I think it would be better to define a type alias (e.g. `JaxObject` or `JaxifiedObject` for this union

egparedes commented on src/jace/util/traits.py:
```diff
@@ -0,0 +1,137 @@
+) -> bool:
+    """Tests if `obj` is a scalar."""
+    # These are the type known to DaCe; Taken from `dace.dtypes`.
+    known_types = {
```
Can't this set be imported from `dace.dtypes` instead of copy pasting their members here?

philip-paul-mueller replied:
No, such a set/list does not exists.

egparedes commented on src/jace/util/util.py:
```diff
@@ -0,0 +1,36 @@
+# JaCe - JAX Just-In-Time compilation using DaCe (Data Centric Parallel Programming)
```
The name of the module is not descriptive enough. I would personally merge it together with the `traits` modules (whose name I also find misleading in a python context) and rename it `definitions` or something like that.

philip-paul-mueller replied:
I do not think so. However, I see where you are coming from, so I renamed the `util.py` to `misc.py`.

egparedes reviewed (changes requested):
This is the missing review for the `util` subpackage.

egparedes reviewed (commented):
Please, make sure docstrings follow our coding conventions (https://google.github.io/styleguide/pyguide.html) and also try to align them to standard conventions in the community (https://peps.python.org/pep-0257/).

In particular:
- Remove the extra blank space after the colon in the description of arguments, only one space is allowed (https://google.github.io/styleguide/pyguide.html#doc-function-args).
- Our coding conventions state that docstring lines shouldn't be longer than 80 chars since they are considered text and not source code. Given that not automatic formatting tools takes care of this, I'm fine with having some lines with slightly more characters here and there, but whole paragraphs with lines around 100 chars are hard to read and should be avoided.
- Follow a consistent style in all docstrings: either descriptive or imperative. I personally think that imperative is slightly more common and  clear, but both are fine so it's up to you to decide.
- Make sure that the docstring content does not mention implementation details which are not relevant for the caller (those should be source code comments).
- Contrary to what I said in previous reviews, I realized that is indeed correct to document `__init__` arguments in the `__init__` docstring instead of the class docstring as long as the style is consistent in the codebase, so feel free to move them to `__init__` if you prefer.
- In multi-line docstrings _"the summary line may be on the same line as the opening quotes or on the next line"_ (PEP 257), and there is a point in our coding guidelines (https://github.com/GridTools/jace/blob/main/CODING_GUIDELINES.md#code-style) explicitly stating that this is the preferred style. We should either honor that point or delete it from the guidelines.

philip-paul-mueller commented:
I generally limited the docstrings to 80 col, however, I do not understand why "is considered text" is not a real explanation and could also be an argument for "no limit".
With the exception where it does not made any sense or the first line string, which should be in my view an exception of the rule anyway.
However, in my opinion our real problem is that we only have partial automation, which is also in my view is much, much worse than having no automation.

I also modified the CODING_GUIDE, as you suggested I removed that with the multi line thing, I was also not able to find it in GT4Py.

egparedes reviewed (approved):
It looks good for now. Let's merge it and keep working in adding features and improving the code.
